### PR TITLE
impl(common): introduce an abstraction for a Cloud location

### PR DESCRIPTION
**Pull request by devbww**
Introduce `google::cloud::Location` to model a location resource like `"projects/project-id/locations/location-id"`, and use this in places where we previously worked directly with strings.

After this change has made it into a release, we can update the quickstart programs (and the scaffolding code that generates them) to use `Location`.

Part of #12631.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12659)

### CODE REVIEW

File: google/cloud/internal/oauth2_external_account_credentials_test.cc [hunk 1]
```diff
@@ -150,13 +152,14 @@ TEST(ExternalAccount, ParseAwsSuccess) {
       "https://sts.{region}.aws.example.com"
       "?Action=GetCallerIdentity&Version=2011-06-15";
   auto constexpr kTestTokenType = "urn:ietf:params:aws:token-type:aws4_request";
-  auto constexpr kTestAudience =
-      "//iam.googleapis.com/projects/$PROJECT_NUMBER/locations/global/"
-      "workloadIdentityPools/$POOL_ID/providers/$PROVIDER_ID";
+  auto const k_test_audience =
+      std::string{"//iam.googleapis.com/" +
+                  Location(Project("$PROJECT_NUMBER"), "global").FullName() +
```
Comment (coryan) [inline google/cloud/internal/oauth2_external_account_credentials_test.cc:157]: I am not sure this is a good application of the new class.  The URL path and resource names match in this case, but they don't have to match (e.g. storage uses`/b/$BUCKET/o/$OBJECT` for the URL path and `projects/_/buckets/$BUCKET/objects/$OBJECT` for the resource name).

Also, the leading `k_` is no longer applicable.

File: examples/batch_logging.cc [hunk 2]
```diff
@@ -45,16 +46,16 @@ void JobLogs(std::vector<std::string> const& argv) {
...
+    auto const name = location.FullName() + "/jobs/" + job_id;
     auto batch = google::cloud::batch_v1::BatchServiceClient(
         google::cloud::batch_v1::MakeBatchServiceConnection());
     auto job = batch.GetJob(name);
     if (!job) throw std::move(job).status();
 
     auto logging = google::cloud::logging_v2::LoggingServiceV2Client(
         google::cloud::logging_v2::MakeLoggingServiceV2Connection());
-    auto const project = google::cloud::Project(project_id);
     auto const log_name = "projects/" + project_id + "/logs/batch_task_logs";
```
Comment (coryan) [inline examples/batch_logging.cc:59]: Should we use the `project.FullName()` here too?

**Review by coryan — CHANGES_REQUESTED** [review]

File: google/cloud/internal/oauth2_external_account_credentials_test.cc (hunk 1 above)
Comment (devbww) [reply google/cloud/internal/oauth2_external_account_credentials_test.cc]: Reverted.

File: examples/batch_logging.cc (hunk 2 above)
Comment (devbww) [reply examples/batch_logging.cc]: Done.

**Review by coryan — APPROVED** [review]
I fixed a typo in the PR title.  You may need to do the same in the commit message.

**Pull request merged**